Add OAuth discovery and SA annotations docs

ahardin-rh opened the pull request:
continuation of https://github.com/openshift/openshift-docs/pull/3092

Preview build: http://file.rdu.redhat.com/~ahardin/11022016/enj-3092/architecture/additional_concepts/authentication.html

enj commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
+
+Any combination of the above syntax can be combined using the following format:
+
+`<scheme>://<hostname><:port>/<path>`
```
`<scheme:>//<hostname><:port>/<path>` (`<scheme:>` instead of `<scheme>:`)

enj commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
+----
+
+<1> `kind` refers to the type of the object being referenced. Currently, only `route` is supported.
+<2> `name` refers to the name of the object. The object must be in the same name space as the service account.
```
name space -> `namespace`

enj commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
+
+[NOTE]
+====
+Not specifying a host name override will replace the host name data from the
```
This is backwards.  If you specify a host name override then you probably made a mistake.

enj commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
+`serviceaccounts.openshift.io/oauth-redirectreference.` such as:
+
+----
+'serviceaccounts.openshift.io/oauth-redirecturi._<name>_'
```
Italics don't work here?

ahardin-rh commented:
@enj Updated. PTAL.

enj commented:
@ahardin-rh LGTM. @mfojtik any comments?

gabemontero commented:
Not a "stop the merge" comment, but it would be cool if the [Jenkins oauth doc ](https://github.com/openshift/openshift-docs/blob/master/using_images/other_images/jenkins.adoc#product-title-oauth-authentication) could be updated to have a link to this new content

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
 $ oc sa get-token <serviceaccount name>
 ----
 
-* `redirect_uri` must match an annotation on the service account that has the
```
A few lines up (currently lines 197 - 202) there are some <replaceables> that have a white space instead of underscore. Like <serviceaccount namespace>, which could be <serviceaccount_namespace> instead.

adellape replied:
Also up above out of this PR's view, a general comment: all the subsections in the "OAuth" section have been just bold headings for a long time. AFAICT I'm the one that did that long ago, but at any rate, they could probably be real subheadings now, so that they show up in TOC / are easier to link to. Just, while you're in this area already!

adellape replied:
There's a couple `<any role>`s above too.

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
+    "name": "jenkins"
+  }
+}
+
```
Normally we want to show YAML instead of JSON for examples, but I think this is fine considering the context given by the stuff before it. :+1:

enj replied:
Actually I misspoke - I used the universal decoder so the server will parse JSON or YAML - it is just really hard to write correct YAML as one giant single line serialized string since whitespace is significant.

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -248,6 +371,79 @@ WWW-Authenticate challenges, users can visit `_<master>_/oauth/token/request`
+
+Applications running in {product-title} may need to discover information about
+the built-in OAuth server. For example, they may need to discover what the
+address of the `_<master>_` server is without manual configuration.  To aid in
```
Drop markup around "master" here?

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -206,13 +206,136 @@ When using a service account as an OAuth client:
 `serviceaccounts.openshift.io/oauth-want-challenges` annotation on the service
 account to *true*.
 
+* `redirect_uri` must match an annotation on the service account:
```
It looks like the bulk of this new content (from here through to *OAuth Server Metadata*) is about this bullet, is that right? I'm assuming that because of the colon. It seems like a lot of info to have trying to be part of a bullet (currently it doesn't have `+`s so it's not really part of the bullet, if that was the intention).

What about making a new heading here... Either just bold or a real `====` subheading within "Service Accounts as OAuth Clients"? Maybe called "Redirect URIs" or "Redirect URIs in Service Accounts".

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -248,6 +371,79 @@ WWW-Authenticate challenges, users can visit `_<master>_/oauth/token/request`
+    "plain",
+    "S256"
+  ]
+}
```
JSON also seems fine here given the context.

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -248,6 +371,79 @@ WWW-Authenticate challenges, users can visit `_<master>_/oauth/token/request`
+Code Challenge Methods* registry.  See
+link:http://www.iana.org/assignments/oauth-parameters[IANA OAuth Parameters].
+
+====
```
Can we lose the `====` block?

ahardin-rh commented:
@enj @gabemontero @adellape Thank you. I addressed all of your comments. Please let me know if I missed anything.

Updated preview builds: 
http://file.rdu.redhat.com/~ahardin/11032016/enj-3092/architecture/additional_concepts/authentication.html

http://file.rdu.redhat.com/~ahardin/11032016/enj-3092/using_images/other_images/jenkins.html#jenkins-configuration-and-usage

adellape commented on architecture/additional_concepts/authentication.adoc:
```diff
@@ -248,8 +374,78 @@ WWW-Authenticate challenges, users can visit `_<master>_/oauth/token/request`
+
+Applications running in {product-title} may need to discover information about
+the built-in OAuth server. For example, they may need to discover what the
+address of the _<master>_ server is without manual configuration.  To aid in
```
Missing backticks here.

ahardin-rh commented:
Thanks! All updated. Merging now.

ahardin-rh commented:
[rev_history]
|xref:../architecture/additional_concepts/authentication.adoc#architecture-additional-concepts-authentication[Additional Concepts -> Authentication]
|Added new xref:../architecture/additional_concepts/authentication.adoc#service-accounts-as-oauth-clients[Service Accounts as OAuth Clients] and xref:../architecture/additional_concepts/authentication.adoc#oauth-server-metadata[OAuth Server Metadata] sections.
%
